typed-racket-test/unit-tests/typecheck-tests.rkt

```diff
@@ -4467,6 +4467,9 @@
         [tc-e/t (let: ([x : (Un Flonum Natural) 0.0])
                   (if (not (natural? x)) x 1.0))
                -Flonum]
+        ;; props + evaluation order
+        [tc-e ((lambda: ((x : (Un Symbol Natural))) (+ (assert x integer?) x)) 1) -Nat]
+        [tc-e ((lambda: ((x : (Un Symbol Natural))) ((begin (assert x integer?) +) x x)) 1) -Nat]
        )
```

Comment:
I think we need more tests. What about

1. Tests where the combinations of the first several arguments then allows the final argument(s) to type check?
2. Tests were at some point the environment becomes absurd (and thus the final arguments are dead code and don't need to be type checked)?

Maybe there's other's that would be useful.

Reply:
update: testing has led to some internal typechecker errors wrt the optimizer, will fix